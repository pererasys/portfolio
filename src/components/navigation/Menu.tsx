/**
 * @author Andrew Perera
 * Copyright (C) 2021
 */

import React, { useEffect, useState } from "react";
import { useRouter } from "next/router";
import Link from "next/link";

import { Squash as Hamburger } from "hamburger-react";
import { enableBodyScroll, disableBodyScroll } from "body-scroll-lock";

import styles from "../../styles/navigation/Menu.module.scss";

interface INavItem {
  label: string;
  path: string;
}

export const NavItem = ({ label, path }: INavItem) => {
  const router = useRouter();

  const isActive = router.pathname === path;

  let classes = [styles.navItem];
  if (isActive) classes.push(styles.active);

  return (
    <li>
      <Link href={path}>
        <a className={classes.join(" ")}>
          <p className={styles.label}>{label}</p>
        </a>
      </Link>
    </li>
  );
};

const ROUTES = [
  { label: "Home", path: "/" },
  { label: "Projects", path: "/projects" },
  { label: "About", path: "/about" },
];

const Menu = () => {
  const [open, setOpen] = useState(false);

  useEffect(() => {
    const elem = document.querySelector("body");
    if (open) {
      disableBodyScroll(elem);
    } else {
      enableBodyScroll(elem);
    }
  }, [open]);

  const classes = [styles.root];

  if (open) classes.push(styles.active);

  return (
    <React.Fragment>
      <div className={styles.action}>
        <Hamburger size={28} color="black" toggled={open} toggle={setOpen} />
      </div>
      <div className={classes.join(" ")}>
        <ul className={styles.navigation}>
          {ROUTES.map((route) => (
            <NavItem {...route} key={route.label} />
          ))}
        </ul>
      </div>
    </React.Fragment>
  );
};

export default Menu;
