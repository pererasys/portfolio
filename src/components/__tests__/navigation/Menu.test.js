import * as nextRouter from "next/router";
import { create, act } from "react-test-renderer";

import { Squash as Hamburger } from "hamburger-react";

import Menu from "../../navigation/Menu";

nextRouter.useRouter = jest.fn();
nextRouter.useRouter.mockImplementation(() => ({
  pathname: "",
  push: jest.fn(),
}));

describe("<Menu />", () => {
  const setup = () => {
    const component = create(<Menu />);

    return {
      component,
    };
  };

  it("should render without crashing", () => {
    const { component } = setup();

    expect(component.toJSON()).toMatchSnapshot();
  });
});
