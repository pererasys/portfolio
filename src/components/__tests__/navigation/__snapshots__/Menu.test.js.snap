// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Menu /> should render without crashing 1`] = `
Array [
  <div>
    <div
      className="hamburger-react"
      onClick={[Function]}
      onKeyUp={[Function]}
      role="button"
      style={
        Object {
          "cursor": "pointer",
          "height": "48px",
          "outline": "none",
          "position": "relative",
          "transition": "0.4s cubic-bezier(0, 0, 0, 1)",
          "userSelect": "none",
          "width": "48px",
        }
      }
      tabIndex={0}
    >
      <div
        style={
          Object {
            "transform": "none",
            "transition": "0.2s cubic-bezier(0, 0, 0, 1) 0.2s",
          }
        }
      >
        <div
          style={
            Object {
              "background": "black",
              "height": "2px",
              "left": "10px",
              "position": "absolute",
              "top": "15px",
              "transform": "none",
              "transition": "0.2s cubic-bezier(0, 0, 0, 1) 0s",
              "width": "28px",
            }
          }
        />
      </div>
      <div
        style={
          Object {
            "opacity": "1",
            "transition": "0.2s cubic-bezier(0, 0, 0, 1)",
          }
        }
      >
        <div
          style={
            Object {
              "background": "black",
              "height": "2px",
              "left": "10px",
              "position": "absolute",
              "top": "23px",
              "transition": "0.2s cubic-bezier(0, 0, 0, 1)",
              "width": "28px",
            }
          }
        />
      </div>
      <div
        style={
          Object {
            "transform": "none",
            "transition": "0.2s cubic-bezier(0, 0, 0, 1) 0.2s",
          }
        }
      >
        <div
          style={
            Object {
              "background": "black",
              "height": "2px",
              "left": "10px",
              "position": "absolute",
              "top": "31px",
              "transform": "none",
              "transition": "0.2s cubic-bezier(0, 0, 0, 1) 0s",
              "width": "28px",
            }
          }
        />
      </div>
    </div>
  </div>,
  <div
    className=""
  >
    <ul>
      <li>
        <a
          className=""
          href="/"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          <p>
            Home
          </p>
        </a>
      </li>
      <li>
        <a
          className=""
          href="/projects"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          <p>
            Projects
          </p>
        </a>
      </li>
      <li>
        <a
          className=""
          href="/about"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          <p>
            About
          </p>
        </a>
      </li>
    </ul>
  </div>,
]
`;
